gdal/port/cpl_userfaultfd.cpp

```diff
@@ -132,44 +149,144 @@ static void * fault_handler(void * ptr)
     uintptr_t fault_addr;
     uint64_t offset;
     off_t bytes_needed;
+    ssize_t bytes_read;
 
     // Poll for event
     if (poll(&pollfd, 1, 16) == -1) break; // 60Hz when no demand
     if ((pollfd.revents & POLLERR) || (pollfd.revents & POLLNVAL)) break;
     if (!(pollfd.revents & POLLIN)) continue;
 
-    // Read page fault event
-    if (read(ctx->uffd, &msg, sizeof(msg)) < 1) {
+    // Read page fault events
+    bytes_read = static_cast<ssize_t>(read(ctx->uffd, ctx->uffd_msgs, MAX_MESSAGES*sizeof(uffd_msg)));
+    if (bytes_read < 1) {
       if (errno == EWOULDBLOCK) continue;
       else break;
     }
 
-    fault_addr = msg.arg.pagefault.address & ~(ctx->page_size-1);
-    offset = static_cast<uint64_t>(fault_addr) - reinterpret_cast<uint64_t>(ctx->vma_ptr);
-    bytes_needed = static_cast<off_t>(ctx->file_size - offset);
-    if (bytes_needed > ctx->page_size) bytes_needed = ctx->page_size;
-
-    // Limit number of pages
-    if ((page_limit > 0) && (pages_used >= page_limit)) {
-      pages_used = 0;
-      ctx->page_ptr = mmap(ctx->page_ptr, ctx->page_size, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS|MAP_FIXED, -1, 0);
-      if (ctx->page_ptr == BAD_MMAP) {
-        ctx->page_ptr = nullptr;
-        break;
+    // If too many pages are in use, evict all pages (evict them from
+    // RAM and swap, not just to swap).  It is impossible to control
+    // which/when threads access the VMA, so access to the VMA has to
+    // forbidden while the activity is in progress.
+    //
+    // That is done by (1) installing special handlers for SIGSEGV and
+    // SIGBUS, (2) mprotecting the VMA so that any threads accessing
+    // it receive either SIGSEGV or SIGBUS (which one is apparently a
+    // function of the C library, at least on one non-Linux GNU
+    // system[1]), (3) unregistering the VMA from userfaultfd,
+    // remapping the VMA to evict the pages, registering the VMA
+    // again, (4) making the VMA accessible again, and finally (5)
+    // restoring the previous signal-handling behavior.
+    //
+    // [1] https://lists.debian.org/debian-bsd/2011/05/msg00032.html
+    pthread_mutex_lock(&mutex);
```

Comment:
the mutex should be unlocked in all the below error break; situations. An alternative would be to use a
static std::mutex mutex;
std::lock_guard<std::mutex> oLock(mutex); 
in a scope,
to avoid having to deal with explicit mutex unlocking

Reply:
I pushed some changes before I saw this comment.  I will wait for the CI to finish then make this change.

Reply:
Done.